File: electron/device-id.js
/** Device fingerprint — must match lib/device_id.py logic on Windows. */
const crypto = require("crypto");
const os = require("os");
const { execSync } = require("child_process");

function run(cmd) {
  try {
    return execSync(cmd, {
      encoding: "utf8",
      timeout: 4000,
      windowsHide: true,
      stdio: ["ignore", "pipe", "ignore"],
    }).trim();
  } catch {
    return "";
  }
}

function windowsParts() {
  const parts = [];
  // 勿用 powershell：部分客户机上 Get-* 会挂起；reg/wmic 更稳
  const guid = run(
    'reg query "HKLM\\SOFTWARE\\Microsoft\\Cryptography" /v MachineGuid'
  );
  const gm = guid.match(/MachineGuid\s+REG_SZ\s+(\S+)/i);
  if (gm) parts.push(gm[1]);

  const board = run("wmic baseboard get serialnumber");
  for (const line of board.split(/\r?\n/)) {
    const t = line.trim();
    if (t && t.toLowerCase() !== "serialnumber") {
      parts.push(t);
      break;
    }
  }
  const vol = run("wmic logicaldisk where DeviceID='C:' get VolumeSerialNumber");
  for (const line of vol.split(/\r?\n/)) {
    const t = line.trim();
    if (t && t.toLowerCase() !== "volumeserialnumber") {
      parts.push(t);
      break;
    }
  }
  parts.push(os.hostname());
  return parts;
}

function computeDeviceId() {
  const env = process.env.MYKNOWLEDGE_DEVICE_ID;
  if (env && env.trim()) return env.trim().slice(0, 64);
  let parts;
  if (process.platform === "win32") {
    parts = windowsParts();
  } else {
    parts = [os.hostname()];
    try {
      const fs = require("fs");
      if (fs.existsSync("/etc/machine-id")) {
        parts.push(fs.readFileSync("/etc/machine-id", "utf8").trim());
      }
    } catch {
      /* ignore */
    }
  }
  const raw = parts.filter(Boolean).join("|") || os.platform();
  return crypto.createHash("sha256").update(raw, "utf8").digest("hex").slice(0, 32);
}

module.exports = { computeDeviceId };

File: lib/device_id.py
"""Stable device fingerprint (SHA-256 hex, first 32 chars)."""

from __future__ import annotations

import hashlib
import os
import platform
import subprocess
import sys
from pathlib import Path


def _run(cmd: list[str]) -> str:
    try:
        out = subprocess.check_output(cmd, stderr=subprocess.DEVNULL, text=True, timeout=5)
        return out.strip()
    except (subprocess.SubprocessError, OSError, FileNotFoundError):
        return ""


def _windows_parts() -> list[str]:
    parts: list[str] = []
    guid = _run(
        [
            "powershell",
            "-NoProfile",
            "-Command",
            "(Get-ItemProperty 'HKLM:\\SOFTWARE\\Microsoft\\Cryptography' -Name MachineGuid).MachineGuid",
        ]
    )
    if guid:
        parts.append(guid)
    board = _run(["wmic", "baseboard", "get", "serialnumber"])
    for line in board.splitlines():
        line = line.strip()
        if line and line.lower() != "serialnumber":
            parts.append(line)
            break
    vol = _run(["wmic", "logicaldisk", "where", "DeviceID='C:'", "get", "VolumeSerialNumber"])
    for line in vol.splitlines():
        line = line.strip()
        if line and line.lower() != "volumeserialnumber":
            parts.append(line)
            break
    parts.append(platform.node())
    return parts


def _unix_parts() -> list[str]:
    parts: list[str] = [platform.node()]
    mid = Path("/etc/machine-id")
    if mid.is_file():
        parts.append(mid.read_text(encoding="utf-8").strip())
    return parts


def compute_device_id() -> str:
    env = os.environ.get("MYKNOWLEDGE_DEVICE_ID", "").strip()
    if env:
        return env[:64]
    if sys.platform == "win32":
        parts = _windows_parts()
    else:
        parts = _unix_parts()
    raw = "|".join(p for p in parts if p) or platform.platform()
    return hashlib.sha256(raw.encode("utf-8")).hexdigest()[:32]
